src/charm.py

```diff
@@ -1,34 +1,146 @@
+        Returns:
+            None: None
+        """
         self.unit.status = MaintenanceStatus("installing MongoDB")
         self._add_mongodb_org_repository()
         self._install_apt_packages(["mongodb-org"])
-        self.unit.status = WaitingStatus("Waiting to start MongoDB")
+        try:
+            mongo_path = subprocess.check_output(["which", "mongod"]).decode().strip()
+            if not mongo_path == "/usr/bin/mongod":
+                logger.error("wrong mongodb path found %s want /usr/bin/mongod", mongo_path)
+                self.unit.status = BlockedStatus("wrong mongod path")
+                return
+        except subprocess.CalledProcessError as e:
+            logger.error("failed to verify mongod installation %s", e)
+            self.unit.status = BlockedStatus("failed to verify mongod installation")
+            return
+
+    def _on_config_changed(self, _: ops.charm.ConfigChangedEvent) -> None:
+        """Event handler for configuration changed events.
+        Args:
+            _ (ops.charm.ConfigChangedEvent): The configuration changed event
+        Returns:
+            None: None
+        """
+        # TODO
+        # - update existing mongo configurations based on user preferences
+        # - add additional configurations as according to spec doc
+        with open("/etc/mongod.conf") as mongo_config_file:
+            mongo_config = yaml.safe_load(mongo_config_file)
+
+        machineIP = str(self.model.get_binding(PEER).network.bind_address)
+        bindIPs = "localhost," + machineIP
+        mongo_config["net"]["bindIp"] = bindIPs
+        if "replication" not in mongo_config:
+            mongo_config["replication"] = {}
+
+        mongo_config["replication"]["replSetName"] = "rs0"
+        with open("/etc/mongod.conf", "w") as mongo_config_file:
+            yaml.dump(mongo_config, mongo_config_file)
+
+        self.unit.status = ops.model.ActiveStatus()
+
+    def _on_start(self, event: ops.charm.StartEvent) -> None:
+        """Enables MongoDB service and initializes replica set
+        Args:
+            event (ops.charm.StartEvent): The triggering start event 
+        Returns:
+            None: None
+        """
+        logger.debug("Running on_start")
+        if not self.unit.is_leader():
+            return
+
+        # start mongo service
+        self._open_port_tcp(self.port)
+        self.unit.status = MaintenanceStatus("enabling MongoDB")
+        try:
+            logger.debug("enabling mongodb")
+            service_resume("mongod.service")
+            service_restart("mongod.service")
+        except subprocess.CalledProcessError as e:
+            logger.error("failed to enable mongo error: %s", e)
+            self.unit.status = BlockedStatus("failed to enable mongo")
+            return
+
+        if not self.mongo.is_ready():
+            self.unit.status = WaitingStatus("Waiting for MongoDB Service")
+            event.defer()
```

Comment:
Hmm, if you're going to do this, then in the section above where you resume/restart, you might want to gate based on whether the service is already running? Right now your flow is:

- on_start: runs mongod, defers because mongo is not yet up
- on_start: triggered again after defer, restarts mongo, then checks if it's up again...